diff --git a/libusb/libusb.h b/libusb/libusb.h
index 58b406f..1d0dd7d 100644
--- a/libusb/libusb.h
+++ b/libusb/libusb.h
@@ -881,6 +881,12 @@ typedef void (LIBUSB_CALL *libusb_transfer_cb_fn)(struct libusb_transfer *transf
  * completed, the library populates the transfer with the results and passes
  * it back to the user.
  */
+
+#if defined(OS_WIN)
+#pragma warning(push)
+#pragma warning(disable:4200)
+#endif  // defined(OS_WIN)
+
 struct libusb_transfer {
 	/** Handle of the device that this transfer will be submitted to */
 	libusb_device_handle *dev_handle;
@@ -939,6 +945,10 @@ struct libusb_transfer {
 	;
 };
 
+#if defined(OS_WIN)
+#pragma warning(pop)
+#endif  // defined(OS_WIN)
+
 /** \ingroup misc
  * Capabilities supported by this instance of libusb. Test if the loaded
  * library supports a given capability by calling
diff --git a/libusb/os/windows_usb.h b/libusb/os/windows_usb.h
index ddbd680..7c2fae5 100644
--- a/libusb/os/windows_usb.h
+++ b/libusb/os/windows_usb.h
@@ -101,7 +101,7 @@ const GUID GUID_DEVINTERFACE_USB_DEVICE = { 0xA5DCBF10, 0x6530, 0x11D2, {0x90, 0
 #if !defined(GUID_DEVINTERFACE_USB_HUB)
 const GUID GUID_DEVINTERFACE_USB_HUB = { 0xF18A0E88, 0xC30C, 0x11D0, {0x88, 0x15, 0x00, 0xA0, 0xC9, 0x06, 0xBE, 0xD8} };
 #endif
-const GUID GUID_NULL = { 0x00000000, 0x0000, 0x0000, {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00} };
+static const GUID GUID_NULL = { 0x00000000, 0x0000, 0x0000, {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00} };
 
 
 /*
