diff -ru a/lib/Target/ARM/ARMSFIPlacement.cpp b/lib/Target/ARM/ARMSFIPlacement.cpp
--- a/lib/Target/ARM/ARMSFIPlacement.cpp	2010-06-03 15:58:33.000000000 -0400
+++ b/lib/Target/ARM/ARMSFIPlacement.cpp	2010-06-03 15:59:54.000000000 -0400
@@ -68,6 +68,8 @@
 }
 
 bool ARMSFIPlacement::TryPredicating(MachineInstr &MI, ARMCC::CondCodes Pred) {
+  return false;
+
   // Can't predicate if it's already predicated.
   // TODO(cbiffle): actually we can, if the conditions match.
   if (TII->isPredicated(&MI)) return false;
Only in b/lib/Target/ARM: ARMSFIPlacement.cpp~
