diff -uNr newlib-1.18.0.orig/libgloss/libnosys/Makefile.in newlib-1.18.0/libgloss/libnosys/Makefile.in
--- newlib-1.18.0.orig/libgloss/libnosys/Makefile.in	2005-12-01 00:43:57.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/Makefile.in	2010-04-11 17:36:16.000000000 +0200
@@ -65,9 +65,15 @@
 	else t='$(program_transform_name)'; echo objcopy | sed -e $$t ; fi`
 
 # object files needed
-OBJS = chown.o close.o environ.o errno.o execve.o fork.o fstat.o \
-	getpid.o gettod.o isatty.o kill.o link.o lseek.o open.o \
-	read.o readlink.o sbrk.o stat.o symlink.o times.o unlink.o \
+OBJS = access.o chdir.o chmod.o chown.o close.o dup.o dup2.o endpwent.o \
+	environ.o errno.o errno_func.o execve.o fchdir.o fchmod.o fchown.o \
+	fcntl.o fork.o fstat.o ftruncate.o get_current_dir_name.o getcwd.o \
+	getegid.o geteuid.o getgid.o getpid.o getppid.o getpwent.o getpwnam.o \
+	getpwnam_r.o getpwuid.o getpwuid_r.o getrusage.o getwd.o gettod.o \
+	getuid.o ioctl.o isatty.o kill.o lchown.o link.o lseek.o mkdir.o \
+	open.o pipe.o pselect.o read.o readlink.o rmdir.o sbrk.o select.o \
+	setegid.o seteuid.o setgid.o setpwent.o setuid.o stat.o symlink.o \
+	times.o truncate.o ttyname.o umask.o unlink.o utime.o utimes.o vfork.o \
 	wait.o write.o _exit.o
 
 # Object files specific to particular targets.
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/access.c newlib-1.18.0/libgloss/libnosys/access.c
--- newlib-1.18.0.orig/libgloss/libnosys/access.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/access.c	2010-04-11 12:14:52.000000000 +0200
@@ -0,0 +1,22 @@
+/*
+ * Stub version of access.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_access, (pathname, mode),
+        const char *pathname _AND
+        int mode)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_access)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/chdir.c newlib-1.18.0/libgloss/libnosys/chdir.c
--- newlib-1.18.0.orig/libgloss/libnosys/chdir.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/chdir.c	2010-04-11 12:19:16.000000000 +0200
@@ -0,0 +1,21 @@
+/*
+ * Stub version of chdir.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_chdir, (path),
+        const char *path)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_chdir)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/chmod.c newlib-1.18.0/libgloss/libnosys/chmod.c
--- newlib-1.18.0.orig/libgloss/libnosys/chmod.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/chmod.c	2010-04-11 12:30:12.000000000 +0200
@@ -0,0 +1,23 @@
+/*
+ * Stub version of chmod.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_chmod, (path, mode),
+        const char *path _AND
+        mode_t mode)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_chmod)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/chown.c newlib-1.18.0/libgloss/libnosys/chown.c
--- newlib-1.18.0.orig/libgloss/libnosys/chown.c	2005-12-01 00:43:57.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/chown.c	2010-04-11 17:07:32.000000000 +0200
@@ -8,16 +8,16 @@
 #include <errno.h>
 #include <sys/types.h>
 #undef errno
-extern int errno;
+extern int *__errno _PARAMS ((void));
 #include "warning.h"
 
 int
 _DEFUN (_chown, (path, owner, group),
-        const char *path  _AND
-        uid_t owner _AND
+        const char *path _AND
+        uid_t owner      _AND
         gid_t group)
 {
-  errno = ENOSYS;
+  (*__errno()) = ENOSYS;
   return -1;
 }
 
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/close.c newlib-1.18.0/libgloss/libnosys/close.c
--- newlib-1.18.0.orig/libgloss/libnosys/close.c	2000-03-17 23:48:50.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/close.c	2010-04-11 11:51:19.000000000 +0200
@@ -7,14 +7,14 @@
 #include <_syslist.h>
 #include <errno.h>
 #undef errno
-extern int errno;
+extern int *__errno _PARAMS ((void));
 #include "warning.h"
 
 int
 _DEFUN (_close, (fildes),
         int fildes)
 {
-  errno = ENOSYS;
+  (*__errno()) = ENOSYS;
   return -1;
 }
 
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/dup.c newlib-1.18.0/libgloss/libnosys/dup.c
--- newlib-1.18.0.orig/libgloss/libnosys/dup.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/dup.c	2010-04-11 12:10:01.000000000 +0200
@@ -0,0 +1,21 @@
+/*
+ * Stub version of dup.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_dup, (oldfd),
+        int oldfd)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_dup)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/dup2.c newlib-1.18.0/libgloss/libnosys/dup2.c
--- newlib-1.18.0.orig/libgloss/libnosys/dup2.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/dup2.c	2010-04-11 12:10:06.000000000 +0200
@@ -0,0 +1,22 @@
+/*
+ * Stub version of dup2.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_dup2, (oldfd, newfd),
+        int oldfd _AND
+        int newfd)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_dup2)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/endpwent.c newlib-1.18.0/libgloss/libnosys/endpwent.c
--- newlib-1.18.0.orig/libgloss/libnosys/endpwent.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/endpwent.c	2010-04-11 16:20:27.000000000 +0200
@@ -0,0 +1,20 @@
+/*
+ * Stub version of endpwent.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+void
+_DEFUN (_endpwent, (),
+        _NOARGS)
+{
+  (*__errno()) = ENOSYS;
+}
+
+stub_warning(_endpwent)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/errno_func.c newlib-1.18.0/libgloss/libnosys/errno_func.c
--- newlib-1.18.0.orig/libgloss/libnosys/errno_func.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/errno_func.c	2010-04-11 17:31:53.000000000 +0200
@@ -0,0 +1,18 @@
+/* Supply a definition of __errno if one not already provided.  */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#undef errno
+extern int errno;
+#include "warning.h"
+
+int *
+_DEFUN (__errno, (),
+        _NOARGS)
+{
+  return &errno;
+}
+
+stub_warning(__errno)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/execve.c newlib-1.18.0/libgloss/libnosys/execve.c
--- newlib-1.18.0.orig/libgloss/libnosys/execve.c	2000-03-17 23:48:50.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/execve.c	2010-04-11 11:51:27.000000000 +0200
@@ -7,7 +7,7 @@
 #include <_syslist.h>
 #include <errno.h>
 #undef errno
-extern int errno;
+extern int *__errno _PARAMS ((void));
 #include "warning.h"
 
 int
@@ -16,7 +16,7 @@
         char **argv  _AND
         char **env)
 {
-  errno = ENOSYS;
+  (*__errno()) = ENOSYS;
   return -1;
 }
 
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/fchdir.c newlib-1.18.0/libgloss/libnosys/fchdir.c
--- newlib-1.18.0.orig/libgloss/libnosys/fchdir.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/fchdir.c	2010-04-11 12:20:46.000000000 +0200
@@ -0,0 +1,21 @@
+/*
+ * Stub version of fchdir.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_fchdir, (fd),
+        int fd)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_fchdir)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/fchmod.c newlib-1.18.0/libgloss/libnosys/fchmod.c
--- newlib-1.18.0.orig/libgloss/libnosys/fchmod.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/fchmod.c	2010-04-11 17:33:10.000000000 +0200
@@ -0,0 +1,23 @@
+/*
+ * Stub version of fchmod.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_fchmod, (filedes, mode),
+        int filedes _AND
+        mode_t mode)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_fchmod)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/fchown.c newlib-1.18.0/libgloss/libnosys/fchown.c
--- newlib-1.18.0.orig/libgloss/libnosys/fchown.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/fchown.c	2010-04-11 17:08:03.000000000 +0200
@@ -0,0 +1,24 @@
+/*
+ * Stub version of fchown.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_fchown, (fd, owner, group),
+        int fd      _AND
+        uid_t owner _AND
+        gid_t group)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_fchown)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/fcntl.c newlib-1.18.0/libgloss/libnosys/fcntl.c
--- newlib-1.18.0.orig/libgloss/libnosys/fcntl.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/fcntl.c	2010-04-11 13:14:13.000000000 +0200
@@ -0,0 +1,22 @@
+/*
+ * Stub version of fcntl.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_fcntl, (fd, cmd),
+        int fd  _AND
+        int cmd _DOTS)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_fcntl)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/fork.c newlib-1.18.0/libgloss/libnosys/fork.c
--- newlib-1.18.0.orig/libgloss/libnosys/fork.c	2000-03-17 23:48:50.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/fork.c	2010-04-11 11:51:30.000000000 +0200
@@ -7,14 +7,14 @@
 #include <_syslist.h>
 #include <errno.h>
 #undef errno
-extern int errno;
+extern int *__errno _PARAMS ((void));
 #include "warning.h"
 
 int
 _DEFUN (_fork, (),
         _NOARGS)
 {
-  errno = ENOSYS;
+  (*__errno()) = ENOSYS;
   return -1;
 }
 
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/fstat.c newlib-1.18.0/libgloss/libnosys/fstat.c
--- newlib-1.18.0.orig/libgloss/libnosys/fstat.c	2000-03-17 23:48:50.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/fstat.c	2010-04-11 17:08:30.000000000 +0200
@@ -9,15 +9,15 @@
 #include <sys/stat.h>
 #include <errno.h>
 #undef errno
-extern int errno;
+extern int *__errno _PARAMS ((void));
 #include "warning.h"
 
 int
 _DEFUN (_fstat, (fildes, st),
-        int          fildes _AND
+        int fildes _AND
         struct stat *st)
 {
-  errno = ENOSYS;
+  (*__errno()) = ENOSYS;
   return -1;
 }
 
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/ftruncate.c newlib-1.18.0/libgloss/libnosys/ftruncate.c
--- newlib-1.18.0.orig/libgloss/libnosys/ftruncate.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/ftruncate.c	2010-04-11 17:33:26.000000000 +0200
@@ -0,0 +1,23 @@
+/*
+ * Stub version of ftruncate.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_ftruncate, (fd, length),
+        int fd _AND
+        off_t length)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_ftruncate)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/get_current_dir_name.c newlib-1.18.0/libgloss/libnosys/get_current_dir_name.c
--- newlib-1.18.0.orig/libgloss/libnosys/get_current_dir_name.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/get_current_dir_name.c	2010-04-11 16:32:01.000000000 +0200
@@ -0,0 +1,21 @@
+/*
+ * Stub version of get_current_dir_name.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+char *
+_DEFUN (_get_current_dir_name, (buf),
+        _NOARGS)
+{
+  (*__errno()) = ENOSYS;
+  return NULL;
+}
+
+stub_warning(_get_current_dir_name)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/getcwd.c newlib-1.18.0/libgloss/libnosys/getcwd.c
--- newlib-1.18.0.orig/libgloss/libnosys/getcwd.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/getcwd.c	2010-04-11 17:08:36.000000000 +0200
@@ -0,0 +1,22 @@
+/*
+ * Stub version of getcwd.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+char *
+_DEFUN (_getcwd, (buf, size),
+        char *buf _AND
+        size_t size)
+{
+  (*__errno()) = ENOSYS;
+  return NULL;
+}
+
+stub_warning(_getcwd)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/getegid.c newlib-1.18.0/libgloss/libnosys/getegid.c
--- newlib-1.18.0.orig/libgloss/libnosys/getegid.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/getegid.c	2010-04-11 13:17:59.000000000 +0200
@@ -0,0 +1,22 @@
+/*
+ * Stub version of getegid.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+gid_t
+_DEFUN (_getegid, (),
+        _NOARGS)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_getegid)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/geteuid.c newlib-1.18.0/libgloss/libnosys/geteuid.c
--- newlib-1.18.0.orig/libgloss/libnosys/geteuid.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/geteuid.c	2010-04-11 13:25:43.000000000 +0200
@@ -0,0 +1,22 @@
+/*
+ * Stub version of geteuid.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+uid_t
+_DEFUN (_geteuid, (),
+        _NOARGS)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_geteuid)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/getgid.c newlib-1.18.0/libgloss/libnosys/getgid.c
--- newlib-1.18.0.orig/libgloss/libnosys/getgid.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/getgid.c	2010-04-11 13:20:19.000000000 +0200
@@ -0,0 +1,22 @@
+/*
+ * Stub version of getgid.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+gid_t
+_DEFUN (_getgid, (),
+        _NOARGS)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_getgid)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/getpid.c newlib-1.18.0/libgloss/libnosys/getpid.c
--- newlib-1.18.0.orig/libgloss/libnosys/getpid.c	2000-03-17 23:48:50.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/getpid.c	2010-04-11 13:16:19.000000000 +0200
@@ -6,15 +6,16 @@
 #include <_ansi.h>
 #include <_syslist.h>
 #include <errno.h>
+#include <sys/types.h>
 #undef errno
-extern int errno;
+extern int *__errno _PARAMS ((void));
 #include "warning.h"
 
-int
+pid_t
 _DEFUN (_getpid, (),
         _NOARGS)
 {
-  errno = ENOSYS;
+  (*__errno()) = ENOSYS;
   return -1;
 }
 
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/getppid.c newlib-1.18.0/libgloss/libnosys/getppid.c
--- newlib-1.18.0.orig/libgloss/libnosys/getppid.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/getppid.c	2010-04-11 13:16:37.000000000 +0200
@@ -0,0 +1,22 @@
+/*
+ * Stub version of getppid.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+pid_t
+_DEFUN (_getppid, (),
+        _NOARGS)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_getppid)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/getpwent.c newlib-1.18.0/libgloss/libnosys/getpwent.c
--- newlib-1.18.0.orig/libgloss/libnosys/getpwent.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/getpwent.c	2010-04-11 16:21:23.000000000 +0200
@@ -0,0 +1,22 @@
+/*
+ * Stub version of getpwent.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+struct passwd *
+_DEFUN (_getpwent, (),
+        _NOARGS)
+{
+  (*__errno()) = ENOSYS;
+  return NULL;
+}
+
+stub_warning(_getpwent)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/getpwnam.c newlib-1.18.0/libgloss/libnosys/getpwnam.c
--- newlib-1.18.0.orig/libgloss/libnosys/getpwnam.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/getpwnam.c	2010-04-11 16:19:55.000000000 +0200
@@ -0,0 +1,22 @@
+/*
+ * Stub version of getpwnam.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+struct passwd *
+_DEFUN (_getpwnam, (name),
+        const char *name)
+{
+  (*__errno()) = ENOSYS;
+  return NULL;
+}
+
+stub_warning(_getpwnam)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/getpwnam_r.c newlib-1.18.0/libgloss/libnosys/getpwnam_r.c
--- newlib-1.18.0.orig/libgloss/libnosys/getpwnam_r.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/getpwnam_r.c	2010-04-11 17:09:04.000000000 +0200
@@ -0,0 +1,25 @@
+/*
+ * Stub version of getpwnam_r.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_getpwnam_r, (name, pwbuf, buf, buflen, pwbufp),
+        const char *name     _AND
+        struct passwd *pwbuf _AND
+        char *buf            _AND
+        size_t buflen        _AND
+        struct passwd **pwbufp)
+{
+  return (*__errno()) = ENOSYS;
+}
+
+stub_warning(_getpwnam_r)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/getpwuid.c newlib-1.18.0/libgloss/libnosys/getpwuid.c
--- newlib-1.18.0.orig/libgloss/libnosys/getpwuid.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/getpwuid.c	2010-04-11 16:22:36.000000000 +0200
@@ -0,0 +1,22 @@
+/*
+ * Stub version of getpwuid.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+struct passwd *
+_DEFUN (_getpwuid, (uid),
+        uid_t uid)
+{
+  (*__errno()) = ENOSYS;
+  return NULL;
+}
+
+stub_warning(_getpwuid)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/getpwuid_r.c newlib-1.18.0/libgloss/libnosys/getpwuid_r.c
--- newlib-1.18.0.orig/libgloss/libnosys/getpwuid_r.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/getpwuid_r.c	2010-04-11 17:09:21.000000000 +0200
@@ -0,0 +1,25 @@
+/*
+ * Stub version of getpwuid_r.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_getpwuid_r, (uid, pwbuf, buf, buflen, pwbufp),
+        uid_t uid            _AND
+        struct passwd *pwbuf _AND
+        char *buf            _AND
+        size_t buflen        _AND
+        struct passwd **pwbufp)
+{
+  return (*__errno()) = ENOSYS;
+}
+
+stub_warning(_getpwuid_r)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/getrusage.c newlib-1.18.0/libgloss/libnosys/getrusage.c
--- newlib-1.18.0.orig/libgloss/libnosys/getrusage.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/getrusage.c	2010-04-11 17:09:26.000000000 +0200
@@ -0,0 +1,23 @@
+/*
+ * Stub version of getrusage.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_getrusage, (who, usage),
+        int who _AND
+        struct rusage *usage)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_getrusage)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/gettod.c newlib-1.18.0/libgloss/libnosys/gettod.c
--- newlib-1.18.0.orig/libgloss/libnosys/gettod.c	2007-07-06 19:00:13.000000000 +0200
+++ newlib-1.18.0/libgloss/libnosys/gettod.c	2010-04-11 11:51:40.000000000 +0200
@@ -9,7 +9,7 @@
 #include <sys/times.h>
 #include <errno.h>
 #undef errno
-extern int errno;
+extern int *__errno _PARAMS ((void));
 #include "warning.h"
 
 struct timeval;
@@ -19,7 +19,7 @@
         struct timeval  *ptimeval  _AND
         void *ptimezone)
 {
-  errno = ENOSYS;
+  (*__errno()) = ENOSYS;
   return -1;
 }
 
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/getuid.c newlib-1.18.0/libgloss/libnosys/getuid.c
--- newlib-1.18.0.orig/libgloss/libnosys/getuid.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/getuid.c	2010-04-11 13:26:06.000000000 +0200
@@ -0,0 +1,22 @@
+/*
+ * Stub version of getuid.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+uid_t
+_DEFUN (_getuid, (),
+        _NOARGS)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_getuid)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/getwd.c newlib-1.18.0/libgloss/libnosys/getwd.c
--- newlib-1.18.0.orig/libgloss/libnosys/getwd.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/getwd.c	2010-04-11 16:30:32.000000000 +0200
@@ -0,0 +1,21 @@
+/*
+ * Stub version of getwd.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+char *
+_DEFUN (_getwd, (buf),
+        char *buf)
+{
+  (*__errno()) = ENOSYS;
+  return NULL;
+}
+
+stub_warning(_getwd)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/ioctl.c newlib-1.18.0/libgloss/libnosys/ioctl.c
--- newlib-1.18.0.orig/libgloss/libnosys/ioctl.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/ioctl.c	2010-04-11 16:39:15.000000000 +0200
@@ -0,0 +1,22 @@
+/*
+ * Stub version of ioctl.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_ioctl, (d, request),
+        int d  _AND
+        int request)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_ioctl)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/isatty.c newlib-1.18.0/libgloss/libnosys/isatty.c
--- newlib-1.18.0.orig/libgloss/libnosys/isatty.c	2008-01-03 23:38:30.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/isatty.c	2010-04-11 11:51:43.000000000 +0200
@@ -7,14 +7,14 @@
 #include <_syslist.h>
 #include <errno.h>
 #undef errno
-extern int errno;
+extern int *__errno _PARAMS ((void));
 #include "warning.h"
 
 int
 _DEFUN (_isatty, (file),
         int file)
 {
-  errno = ENOSYS;
+  (*__errno()) = ENOSYS;
   return 0;
 }
 
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/kill.c newlib-1.18.0/libgloss/libnosys/kill.c
--- newlib-1.18.0.orig/libgloss/libnosys/kill.c	2000-03-17 23:48:50.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/kill.c	2010-04-11 11:51:45.000000000 +0200
@@ -7,7 +7,7 @@
 #include <_syslist.h>
 #include <errno.h>
 #undef errno
-extern int errno;
+extern int *__errno _PARAMS ((void));
 #include "warning.h"
 
 int
@@ -15,7 +15,7 @@
         int pid  _AND
         int sig)
 {
-  errno = ENOSYS;
+  (*__errno()) = ENOSYS;
   return -1;
 }
 
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/lchown.c newlib-1.18.0/libgloss/libnosys/lchown.c
--- newlib-1.18.0.orig/libgloss/libnosys/lchown.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/lchown.c	2010-04-11 17:09:57.000000000 +0200
@@ -0,0 +1,24 @@
+/*
+ * Stub version of lchown.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_lchown, (path, owner, group),
+        const char *path  _AND
+        uid_t owner       _AND
+        gid_t group)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_lchown)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/link.c newlib-1.18.0/libgloss/libnosys/link.c
--- newlib-1.18.0.orig/libgloss/libnosys/link.c	2000-03-17 23:48:50.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/link.c	2010-04-11 11:51:52.000000000 +0200
@@ -7,7 +7,7 @@
 #include <_syslist.h>
 #include <errno.h>
 #undef errno
-extern int errno;
+extern int *__errno _PARAMS ((void));
 #include "warning.h"
 
 int
@@ -15,7 +15,7 @@
         char *existing _AND
         char *new)
 {
-  errno = ENOSYS;
+  (*__errno()) = ENOSYS;
   return -1;
 }
 
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/lseek.c newlib-1.18.0/libgloss/libnosys/lseek.c
--- newlib-1.18.0.orig/libgloss/libnosys/lseek.c	2000-03-17 23:48:50.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/lseek.c	2010-04-11 11:51:55.000000000 +0200
@@ -7,7 +7,7 @@
 #include <_syslist.h>
 #include <errno.h>
 #undef errno
-extern int errno;
+extern int *__errno _PARAMS ((void));
 #include "warning.h"
 
 int
@@ -16,7 +16,7 @@
         int   ptr   _AND
         int   dir)
 {
-  errno = ENOSYS;
+  (*__errno()) = ENOSYS;
   return -1;
 }
 
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/mkdir.c newlib-1.18.0/libgloss/libnosys/mkdir.c
--- newlib-1.18.0.orig/libgloss/libnosys/mkdir.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/mkdir.c	2010-04-11 16:42:11.000000000 +0200
@@ -0,0 +1,23 @@
+/*
+ * Stub version of mkdir.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_mkdir, (pathname, mode),
+        const char *pathname  _AND
+        mode_t mode)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_mkdir)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/open.c newlib-1.18.0/libgloss/libnosys/open.c
--- newlib-1.18.0.orig/libgloss/libnosys/open.c	2000-03-17 23:48:50.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/open.c	2010-04-11 11:51:57.000000000 +0200
@@ -7,7 +7,7 @@
 #include <_syslist.h>
 #include <errno.h>
 #undef errno
-extern int errno;
+extern int *__errno _PARAMS ((void));
 #include "warning.h"
 
 int
@@ -16,7 +16,7 @@
         int   flags _AND
         int   mode)
 {
-  errno = ENOSYS;
+  (*__errno()) = ENOSYS;
   return -1;
 }
 
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/pipe.c newlib-1.18.0/libgloss/libnosys/pipe.c
--- newlib-1.18.0.orig/libgloss/libnosys/pipe.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/pipe.c	2010-04-11 12:10:16.000000000 +0200
@@ -0,0 +1,21 @@
+/*
+ * Stub version of pipe.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_pipe, (fd),
+        int fd[2])
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_pipe)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/pselect.c newlib-1.18.0/libgloss/libnosys/pselect.c
--- newlib-1.18.0.orig/libgloss/libnosys/pselect.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/pselect.c	2010-04-11 17:34:45.000000000 +0200
@@ -0,0 +1,28 @@
+/*
+ * Stub version of pselect.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#include <sys/signal.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_pselect, (n, readfds, writefds, exceptfds, timeout, sigmask),
+        int n                   _AND
+        fd_set *readfds         _AND
+        fd_set *writefds        _AND
+        fd_set *exceptfds       _AND
+        struct timeval *timeout _AND
+        const sigset_t *sigmask)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_pselect)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/read.c newlib-1.18.0/libgloss/libnosys/read.c
--- newlib-1.18.0.orig/libgloss/libnosys/read.c	2000-03-17 23:48:50.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/read.c	2010-04-11 11:52:00.000000000 +0200
@@ -7,7 +7,7 @@
 #include <_syslist.h>
 #include <errno.h>
 #undef errno
-extern int errno;
+extern int *__errno _PARAMS ((void));
 #include "warning.h"
 
 int
@@ -16,7 +16,7 @@
         char *ptr   _AND
         int   len)
 {
-  errno = ENOSYS;
+  (*__errno()) = ENOSYS;
   return -1;
 }
 
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/readlink.c newlib-1.18.0/libgloss/libnosys/readlink.c
--- newlib-1.18.0.orig/libgloss/libnosys/readlink.c	2005-12-01 00:43:57.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/readlink.c	2010-04-11 17:43:05.000000000 +0200
@@ -8,16 +8,16 @@
 #include <errno.h>
 #include <sys/types.h>
 #undef errno
-extern int errno;
+extern int *__errno _PARAMS ((void));
 #include "warning.h"
 
 int
 _DEFUN (_readlink, (path, buf, bufsize),
         const char *path _AND
-        char *buf _AND
+        char *buf        _AND
         size_t bufsize)
 {
-  errno = ENOSYS;
+  (*__errno()) = ENOSYS;
   return -1;
 }
 
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/rmdir.c newlib-1.18.0/libgloss/libnosys/rmdir.c
--- newlib-1.18.0.orig/libgloss/libnosys/rmdir.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/rmdir.c	2010-04-11 16:42:46.000000000 +0200
@@ -0,0 +1,22 @@
+/*
+ * Stub version of rmdir.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_rmdir, (pathname),
+        const char *pathname)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_rmdir)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/select.c newlib-1.18.0/libgloss/libnosys/select.c
--- newlib-1.18.0.orig/libgloss/libnosys/select.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/select.c	2010-04-11 17:10:48.000000000 +0200
@@ -0,0 +1,26 @@
+/*
+ * Stub version of select.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_select, (n, readfds, writefds, exceptfds, timeout),
+        int n             _AND
+        fd_set *readfds   _AND
+        fd_set *writefds  _AND
+        fd_set *exceptfds _AND
+        struct timeval *timeout)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_select)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/setegid.c newlib-1.18.0/libgloss/libnosys/setegid.c
--- newlib-1.18.0.orig/libgloss/libnosys/setegid.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/setegid.c	2010-04-11 13:22:46.000000000 +0200
@@ -0,0 +1,22 @@
+/*
+ * Stub version of setegid.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_setegid, (gid),
+        gid_t gid)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_setegid)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/seteuid.c newlib-1.18.0/libgloss/libnosys/seteuid.c
--- newlib-1.18.0.orig/libgloss/libnosys/seteuid.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/seteuid.c	2010-04-11 13:26:36.000000000 +0200
@@ -0,0 +1,22 @@
+/*
+ * Stub version of seteuid.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_seteuid, (uid),
+        uid_t uid)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_seteuid)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/setgid.c newlib-1.18.0/libgloss/libnosys/setgid.c
--- newlib-1.18.0.orig/libgloss/libnosys/setgid.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/setgid.c	2010-04-11 13:23:55.000000000 +0200
@@ -0,0 +1,22 @@
+/*
+ * Stub version of setgid.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_setgid, (gid),
+        gid_t gid)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_setgid)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/setpwent.c newlib-1.18.0/libgloss/libnosys/setpwent.c
--- newlib-1.18.0.orig/libgloss/libnosys/setpwent.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/setpwent.c	2010-04-11 16:24:06.000000000 +0200
@@ -0,0 +1,20 @@
+/*
+ * Stub version of setpwent.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+void
+_DEFUN (_setpwent, (),
+        _NOARGS)
+{
+  (*__errno()) = ENOSYS;
+}
+
+stub_warning(_setpwent)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/setuid.c newlib-1.18.0/libgloss/libnosys/setuid.c
--- newlib-1.18.0.orig/libgloss/libnosys/setuid.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/setuid.c	2010-04-11 13:26:49.000000000 +0200
@@ -0,0 +1,22 @@
+/*
+ * Stub version of setuid.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_setuid, (uid),
+        uid_t uid)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_setuid)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/stat.c newlib-1.18.0/libgloss/libnosys/stat.c
--- newlib-1.18.0.orig/libgloss/libnosys/stat.c	2000-03-17 23:48:50.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/stat.c	2010-04-11 11:52:10.000000000 +0200
@@ -9,7 +9,7 @@
 #include <sys/stat.h>
 #include <errno.h>
 #undef errno
-extern int errno;
+extern int *__errno _PARAMS ((void));
 #include "warning.h"
 
 int
@@ -17,7 +17,7 @@
         const char  *file _AND
         struct stat *st)
 {
-  errno = ENOSYS;
+  (*__errno()) = ENOSYS;
   return -1;
 }
 
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/symlink.c newlib-1.18.0/libgloss/libnosys/symlink.c
--- newlib-1.18.0.orig/libgloss/libnosys/symlink.c	2005-12-01 00:43:57.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/symlink.c	2010-04-11 11:52:13.000000000 +0200
@@ -7,7 +7,7 @@
 #include <_syslist.h>
 #include <errno.h>
 #undef errno
-extern int errno;
+extern int *__errno _PARAMS ((void));
 #include "warning.h"
 
 int
@@ -15,7 +15,7 @@
         const char *path1 _AND
         const char *path2)
 {
-  errno = ENOSYS;
+  (*__errno()) = ENOSYS;
   return -1;
 }
 
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/times.c newlib-1.18.0/libgloss/libnosys/times.c
--- newlib-1.18.0.orig/libgloss/libnosys/times.c	2000-03-17 23:48:50.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/times.c	2010-04-11 11:52:15.000000000 +0200
@@ -8,14 +8,14 @@
 #include <sys/times.h>
 #include <errno.h>
 #undef errno
-extern int errno;
+extern int *__errno _PARAMS ((void));
 #include "warning.h"
 
 clock_t
 _DEFUN (_times, (buf),
         struct tms *buf)
 {
-  errno = ENOSYS;
+  (*__errno()) = ENOSYS;
   return -1;
 }
 
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/truncate.c newlib-1.18.0/libgloss/libnosys/truncate.c
--- newlib-1.18.0.orig/libgloss/libnosys/truncate.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/truncate.c	2010-04-11 17:33:33.000000000 +0200
@@ -0,0 +1,23 @@
+/*
+ * Stub version of truncate.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_truncate, (path, length),
+        const char *path _AND
+        off_t length)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_truncate)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/ttyname.c newlib-1.18.0/libgloss/libnosys/ttyname.c
--- newlib-1.18.0.orig/libgloss/libnosys/ttyname.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/ttyname.c	2010-04-11 17:14:01.000000000 +0200
@@ -0,0 +1,21 @@
+/*
+ * Stub version of ttyname.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+char *
+_DEFUN (_ttyname, (fd),
+        int fd)
+{
+  (*__errno()) = ENOSYS;
+  return NULL;
+}
+
+stub_warning(_ttyname)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/umask.c newlib-1.18.0/libgloss/libnosys/umask.c
--- newlib-1.18.0.orig/libgloss/libnosys/umask.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/umask.c	2010-04-11 17:15:05.000000000 +0200
@@ -0,0 +1,22 @@
+/*
+ * Stub version of umask.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+mode_t
+_DEFUN (_umask, (mode),
+        mode_t mode)
+{
+  (*__errno()) = ENOSYS;
+  return 0777;
+}
+
+stub_warning(_umask)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/unlink.c newlib-1.18.0/libgloss/libnosys/unlink.c
--- newlib-1.18.0.orig/libgloss/libnosys/unlink.c	2000-03-17 23:48:50.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/unlink.c	2010-04-11 11:52:21.000000000 +0200
@@ -7,14 +7,14 @@
 #include <_syslist.h>
 #include <errno.h>
 #undef errno
-extern int errno;
+extern int *__errno _PARAMS ((void));
 #include "warning.h"
 
 int
 _DEFUN (_unlink, (name),
         char *name)
 {
-  errno = ENOSYS;
+  (*__errno()) = ENOSYS;
   return -1;
 }
 
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/utime.c newlib-1.18.0/libgloss/libnosys/utime.c
--- newlib-1.18.0.orig/libgloss/libnosys/utime.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/utime.c	2010-04-11 17:16:19.000000000 +0200
@@ -0,0 +1,23 @@
+/*
+ * Stub version of utime.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/types.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_utime, (filename, buf),
+        const char *filename _AND
+        const struct utimbuf *buf)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_utime)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/utimes.c newlib-1.18.0/libgloss/libnosys/utimes.c
--- newlib-1.18.0.orig/libgloss/libnosys/utimes.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/utimes.c	2010-04-11 17:35:30.000000000 +0200
@@ -0,0 +1,23 @@
+/*
+ * Stub version of utimes.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#include <sys/time.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_utimes, (filename, tv),
+        const char *filename _AND
+        const struct timeval tv[2])
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_utimes)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/vfork.c newlib-1.18.0/libgloss/libnosys/vfork.c
--- newlib-1.18.0.orig/libgloss/libnosys/vfork.c	1970-01-01 01:00:00.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/vfork.c	2010-04-11 12:05:18.000000000 +0200
@@ -0,0 +1,21 @@
+/*
+ * Stub version of vfork.
+ */
+
+#include "config.h"
+#include <_ansi.h>
+#include <_syslist.h>
+#include <errno.h>
+#undef errno
+extern int *__errno _PARAMS ((void));
+#include "warning.h"
+
+int
+_DEFUN (_vfork, (),
+        _NOARGS)
+{
+  (*__errno()) = ENOSYS;
+  return -1;
+}
+
+stub_warning(_vfork)
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/wait.c newlib-1.18.0/libgloss/libnosys/wait.c
--- newlib-1.18.0.orig/libgloss/libnosys/wait.c	2000-03-17 23:48:50.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/wait.c	2010-04-11 11:52:24.000000000 +0200
@@ -7,14 +7,14 @@
 #include <_syslist.h>
 #include <errno.h>
 #undef errno
-extern int errno;
+extern int *__errno _PARAMS ((void));
 #include "warning.h"
 
 int
 _DEFUN (_wait, (status),
         int  *status)
 {
-  errno = ENOSYS;
+  (*__errno()) = ENOSYS;
   return -1;
 }
 
diff -uNr newlib-1.18.0.orig/libgloss/libnosys/write.c newlib-1.18.0/libgloss/libnosys/write.c
--- newlib-1.18.0.orig/libgloss/libnosys/write.c	2000-03-17 23:48:50.000000000 +0100
+++ newlib-1.18.0/libgloss/libnosys/write.c	2010-04-11 11:52:28.000000000 +0200
@@ -7,7 +7,7 @@
 #include <_syslist.h>
 #include <errno.h>
 #undef errno
-extern int errno;
+extern int *__errno _PARAMS ((void));
 #include "warning.h"
 
 int
@@ -16,7 +16,7 @@
         char *ptr   _AND
         int   len)
 {
-  errno = ENOSYS;
+  (*__errno()) = ENOSYS;
   return -1;
 }
 
