Index: source/common/locid.cpp
===================================================================
--- source/common/locid.cpp	(revision 36834)
+++ source/common/locid.cpp	(working copy)
@@ -484,6 +484,13 @@
         /* preset all fields to empty */
         language[0] = script[0] = country[0] = 0;

+        // Need to reset baseName. Otherwise, when a Locale object created with
+        // the default constructor is changed with setFromPOSIXID() later
+        // (e.g. locales obtained with getAvailableLocales()),
+        // baseName will be still that of the default locale instead of one
+        // corresponding to localeID.
+        baseName = NULL;
+
         // "canonicalize" the locale ID to ICU/Java format
         err = U_ZERO_ERROR;
         length = canonicalize ?
