diff -wurp -N orig/libavcodec/Makefile ffmpeg/libavcodec/Makefile
--- orig/libavcodec/Makefile	2011-07-12 20:47:24.792991744 -0700
+++ ffmpeg/libavcodec/Makefile	2011-07-12 20:47:24.882984792 -0700
@@ -556,8 +556,7 @@ OBJS-$(CONFIG_MOV_MUXER)               +
 OBJS-$(CONFIG_MPEGTS_MUXER)            += mpegvideo.o mpeg4audio.o
 OBJS-$(CONFIG_MPEGTS_DEMUXER)          += mpeg4audio.o mpegaudiodata.o
 OBJS-$(CONFIG_NUT_MUXER)               += mpegaudiodata.o
-OBJS-$(CONFIG_OGG_DEMUXER)             += flacdec.o flacdata.o flac.o \
-                                          dirac.o mpeg12data.o vorbis_data.o
+OBJS-$(CONFIG_OGG_DEMUXER)             += dirac.o mpeg12data.o vorbis_data.o
 OBJS-$(CONFIG_OGG_MUXER)               += xiph.o flacdec.o flacdata.o flac.o \
                                           vorbis_data.o
 OBJS-$(CONFIG_RTP_MUXER)               += mpeg4audio.o mpegvideo.o xiph.o
diff -wurp -N orig/libavformat/Makefile ffmpeg/libavformat/Makefile
--- orig/libavformat/Makefile	2011-07-12 20:47:24.824391331 -0700
+++ ffmpeg/libavformat/Makefile	2011-07-12 20:47:24.882984792 -0700
@@ -163,13 +163,13 @@ OBJS-$(CONFIG_NULL_MUXER)               
 OBJS-$(CONFIG_NUT_DEMUXER)               += nutdec.o nut.o riff.o
 OBJS-$(CONFIG_NUT_MUXER)                 += nutenc.o nut.o riff.o
 OBJS-$(CONFIG_NUV_DEMUXER)               += nuv.o riff.o
+OBJS-$(CONFIG_FLAC_DECODER)              += oggparseflac.o
+OBJS-$(CONFIG_DIRAC_DEMUXER)             += oggparsedirac.o
+OBJS-$(CONFIG_LIBSPEEX)                  += oggparsespeex.o
 OBJS-$(CONFIG_OGG_DEMUXER)               += oggdec.o         \
                                             oggparsecelt.o   \
-                                            oggparsedirac.o  \
-                                            oggparseflac.o   \
                                             oggparseogm.o    \
                                             oggparseskeleton.o \
-                                            oggparsespeex.o  \
                                             oggparsetheora.o \
                                             oggparsevorbis.o \
                                             riff.o \
diff -wurp -N orig/libavformat/oggdec.c ffmpeg/libavformat/oggdec.c
--- orig/libavformat/oggdec.c	2011-07-12 20:47:24.824391331 -0700
+++ ffmpeg/libavformat/oggdec.c	2011-07-12 20:47:24.882984792 -0700
@@ -40,14 +40,24 @@
 
 static const struct ogg_codec * const ogg_codecs[] = {
     &ff_skeleton_codec,
+#if CONFIG_DIRAC_DEMUXER
     &ff_dirac_codec,
+#endif
+#if CONFIG_LIBSPEEX
     &ff_speex_codec,
+#endif
     &ff_vorbis_codec,
     &ff_theora_codec,
+#if CONFIG_FLAC_DECODER
     &ff_flac_codec,
+#endif
     &ff_celt_codec,
+#if CONFIG_DIRAC_DEMUXER
     &ff_old_dirac_codec,
+#endif
+#if CONFIG_FLAC_DECODER
     &ff_old_flac_codec,
+#endif
     &ff_ogm_video_codec,
     &ff_ogm_audio_codec,
     &ff_ogm_text_codec,
