diff -wurp -N orig/libavformat/oggdec.c ffmpeg/libavformat/oggdec.c
--- orig/libavformat/oggdec.c   2011-08-24 20:49:56.784219402 -0700
+++ ffmpeg/libavformat/oggdec.c 2011-08-24 20:49:56.832983905 -0700
@@ -480,18 +480,6 @@ static int ogg_get_length(AVFormatContext *s)
 
     ogg_restore (s, 0);
 
-    ogg_save (s);
-    avio_seek (s->pb, 0, SEEK_SET);
-    while (!ogg_read_page (s, &i)){
-        if (ogg->streams[i].granule != -1 && ogg->streams[i].granule != 0 &&
-            ogg->streams[i].codec) {
-            s->streams[i]->duration -=
-                ogg_gptopts (s, i, ogg->streams[i].granule, NULL);
-            break;
-        }
-    }
-    ogg_restore (s, 0);
-
     return 0;
 }
 
