diff -wurp -N orig/libavformat/matroskadec.c patched-ffmpeg/libavformat/matroskadec.c
--- orig/libavformat/matroskadec.c	2011-10-20 11:28:27.935337424 -0700
+++ patched-ffmpeg/libavformat/matroskadec.c	2011-10-20 11:32:03.845401593 -0700
@@ -1241,7 +1241,6 @@ static int matroska_parse_seekhead_entry
 static void matroska_execute_seekhead(MatroskaDemuxContext *matroska)
 {
     EbmlList *seekhead_list = &matroska->seekhead;
-    MatroskaSeekhead *seekhead = seekhead_list->elem;
     int64_t before_pos = avio_tell(matroska->ctx->pb);
     int i;
 
@@ -1251,6 +1250,7 @@ static void matroska_execute_seekhead(Ma
         return;
 
     for (i = 0; i < seekhead_list->nb_elem; i++) {
+        MatroskaSeekhead *seekhead = seekhead_list->elem;
         if (seekhead[i].pos <= before_pos)
             continue;
 
