diff -wurp -N orig/libavcodec/vorbis.c patched-ffmpeg/libavcodec/vorbis.c
--- orig/libavcodec/vorbis.c	2011-10-20 15:33:53.446580551 -0700
+++ patched-ffmpeg/libavcodec/vorbis.c	2011-10-20 15:34:15.605319551 -0700
@@ -150,7 +150,7 @@ void ff_vorbis_ready_floor1_list(vorbis_
     }
 }
 
-static inline void render_line_unrolled(intptr_t x, intptr_t y, int x1,
+static inline void render_line_unrolled(intptr_t x, unsigned char y, int x1,
                                         intptr_t sy, int ady, int adx,
                                         float *buf)
 {
@@ -185,7 +185,7 @@ static void render_line(int x0, int y0, 
     } else {
         int base = dy / adx;
         int x    = x0;
-        int y    = y0;
+        unsigned char y = y0;
         int err  = -adx;
         ady -= FFABS(base) * adx;
         while (++x < x1) {
