diff -wurp -N orig/libavformat/mov.c ffmpeg/libavformat/mov.c
--- orig/libavformat/mov.c	2012-02-03 11:39:21.961119253 -0800
+++ ffmpeg/libavformat/mov.c	2012-02-03 11:39:56.500485313 -0800
@@ -1558,7 +1558,7 @@ static void mov_build_index(MOVContext *
         unsigned int stts_sample = 0;
         unsigned int sample_size;
         unsigned int distance = 0;
-        int key_off = sc->keyframes && sc->keyframes[0] == 1;
+        int key_off = sc->keyframe_count && sc->keyframes[0] == 1;
 
         current_dts -= sc->dts_shift;
 
