diff --git a/libavcodec/vp3.c b/libavcodec/vp3.c
index 5bf2fee..ddfb49f 100644
--- a/libavcodec/vp3.c
+++ b/libavcodec/vp3.c
@@ -1635,7 +1635,7 @@ static av_cold int allocate_tables(AVCodecContext *avctx)
     s->superblock_coding = av_malloc(s->superblock_count);
     s->all_fragments = av_malloc(s->fragment_count * sizeof(Vp3Fragment));
     s->coded_fragment_list[0] = av_malloc(s->fragment_count * sizeof(int));
-    s->dct_tokens_base = av_malloc(64*s->fragment_count * sizeof(*s->dct_tokens_base));
+    s->dct_tokens_base = av_mallocz(64*s->fragment_count * sizeof(*s->dct_tokens_base));
     s->motion_val[0] = av_malloc(y_fragment_count * sizeof(*s->motion_val[0]));
     s->motion_val[1] = av_malloc(c_fragment_count * sizeof(*s->motion_val[1]));
 
