From 510cd79306562b5c690f22adc127b00e8f21f7e8 Mon Sep 17 00:00:00 2001
From: Egor Pasko <pasko@google.com>
Date: Tue, 18 May 2010 15:49:36 +0200
Subject: [PATCH 11/12] [gcc] Make copy-propagated pointers be addressed in Pmode.

BUG: http://code.google.com/p/nativeclient/issues/detail?id=453
---
 gcc/gcc/config/i386/i386.c |    7 ++++++-
 1 files changed, 6 insertions(+), 1 deletions(-)

diff --git a/gcc/gcc/config/i386/i386.c b/gcc/gcc/config/i386/i386.c
index d23b4ba..42c2dbf 100644
--- a/gcc/gcc/config/i386/i386.c
+++ b/gcc/gcc/config/i386/i386.c
@@ -19102,7 +19102,12 @@ ix86_expand_call (rtx retval, rtx fnaddr, rtx callarg1,
 	   ? !sibcall_insn_operand (XEXP (fnaddr, 0), Pmode)
 	   : !call_insn_operand (XEXP (fnaddr, 0), Pmode))
     {
-      fnaddr = copy_to_mode_reg (Pmode, XEXP (fnaddr, 0));
+      fnaddr = copy_to_mode_reg (ptr_mode, XEXP (fnaddr, 0));
+      if (Pmode != ptr_mode)
+	{
+	  fnaddr = gen_lowpart (Pmode, fnaddr);
+	  fnaddr = force_reg(Pmode, fnaddr);
+	}
       fnaddr = gen_rtx_MEM (QImode, fnaddr);
     }
 
-- 
1.7.0.1

