--- a/src/google/protobuf/repeated_field.cc
+++ b/src/google/protobuf/repeated_field.cc
@@ -120,15 +120,5 @@ MessageLite* RepeatedPtrFieldBase::AddWeak(const MessageLite* prototype) {
 
 }  // namespace internal
 
-
-template class PROTOBUF_EXPORT RepeatedField<bool>;
-template class PROTOBUF_EXPORT RepeatedField<int32>;
-template class PROTOBUF_EXPORT RepeatedField<uint32>;
-template class PROTOBUF_EXPORT RepeatedField<int64>;
-template class PROTOBUF_EXPORT RepeatedField<uint64>;
-template class PROTOBUF_EXPORT RepeatedField<float>;
-template class PROTOBUF_EXPORT RepeatedField<double>;
-template class PROTOBUF_EXPORT RepeatedPtrField<std::string>;
-
 }  // namespace protobuf
 }  // namespace google
--- a/src/google/protobuf/repeated_field.h
+++ b/src/google/protobuf/repeated_field.h
@@ -2664,16 +2664,6 @@ UnsafeArenaAllocatedRepeatedPtrFieldBackInserter(
       mutable_field);
 }
 
-// Extern declarations of common instantiations to reduce libray bloat.
-extern template class PROTOBUF_EXPORT RepeatedField<bool>;
-extern template class PROTOBUF_EXPORT RepeatedField<int32>;
-extern template class PROTOBUF_EXPORT RepeatedField<uint32>;
-extern template class PROTOBUF_EXPORT RepeatedField<int64>;
-extern template class PROTOBUF_EXPORT RepeatedField<uint64>;
-extern template class PROTOBUF_EXPORT RepeatedField<float>;
-extern template class PROTOBUF_EXPORT RepeatedField<double>;
-extern template class PROTOBUF_EXPORT RepeatedPtrField<std::string>;
-
 }  // namespace protobuf
 }  // namespace google
 
