diff -pu a/nss/lib/ssl/sslsnce.c b/nss/lib/ssl/sslsnce.c
--- a/nss/lib/ssl/sslsnce.c	2013-07-31 12:07:10.974699609 -0700
+++ b/nss/lib/ssl/sslsnce.c	2013-07-31 14:12:33.185058439 -0700
@@ -86,6 +86,11 @@
 #include "nspr.h"
 #include "sslmutex.h"
 
+/* AES_256_KEY_LENGTH was added to blapit.h in NSS 3.12.10. */
+#ifndef AES_256_KEY_LENGTH
+#define AES_256_KEY_LENGTH      32  /* bytes */
+#endif
+
 /*
 ** Format of a cache entry in the shared memory.
 */ 
