diff --git a/lib/ssl/sslinfo.c b/lib/ssl/sslinfo.c
index 527b1a4..c59879c 100644
--- a/lib/ssl/sslinfo.c
+++ b/lib/ssl/sslinfo.c
@@ -406,8 +406,13 @@ SSL_ExportKeyingMaterial(PRFileDesc *fd,
         return SECFailure;
     }
 
+    ssl_GetRecvBufLock(ss);
+    ssl_GetSSL3HandshakeLock(ss);
+
     if (ss->version < SSL_LIBRARY_VERSION_3_1_TLS) {
         PORT_SetError(SSL_ERROR_FEATURE_NOT_SUPPORTED_FOR_VERSION);
+        ssl_ReleaseSSL3HandshakeLock(ss);
+        ssl_ReleaseRecvBufLock(ss);
         return SECFailure;
     }
 
@@ -418,6 +423,8 @@ SSL_ExportKeyingMaterial(PRFileDesc *fd,
     }
     val = PORT_Alloc(valLen);
     if (!val) {
+        ssl_ReleaseSSL3HandshakeLock(ss);
+        ssl_ReleaseRecvBufLock(ss);
         return SECFailure;
     }
     i = 0;
@@ -445,6 +452,8 @@ SSL_ExportKeyingMaterial(PRFileDesc *fd,
                                          valLen, out, outLen);
     }
     ssl_ReleaseSpecReadLock(ss);
+    ssl_ReleaseSSL3HandshakeLock(ss);
+    ssl_ReleaseRecvBufLock(ss);
 
     PORT_ZFree(val, valLen);
     return rv;
