diff --git a/lib/ssl/ssl3ecc.c b/lib/ssl/ssl3ecc.c
index 94008a0..6d89bbe 100644
--- a/lib/ssl/ssl3ecc.c
+++ b/lib/ssl/ssl3ecc.c
@@ -1093,6 +1093,7 @@ static const PRUint8 ecPtFmt[6] = {
 static PRBool
 ssl3_SuiteBOnly(sslSocket *ss)
 {
+#if 0
     /* See if we can support small curves (like 163). If not, assume we can
      * only support Suite-B curves (P-256, P-384, P-521). */
     PK11SlotInfo *slot =
@@ -1106,6 +1107,9 @@ ssl3_SuiteBOnly(sslSocket *ss)
     /* we can, presume we can do all curves */
     PK11_FreeSlot(slot);
     return PR_FALSE;
+#else
+    return PR_TRUE;
+#endif
 }
 
 /* Send our "canned" (precompiled) Supported Elliptic Curves extension,
