Index: coregrind/m_main.c
===================================================================
--- coregrind/m_main.c	(revision 10461)
+++ coregrind/m_main.c	(working copy)
@@ -2175,11 +2186,11 @@
    if (VG_(clo_track_fds))
       VG_(show_open_fds)();
 
+   VG_TDICT_CALL(tool_fini, 0/*exitcode*/);
+
    if (VG_(needs).core_errors || VG_(needs).tool_errors)
       VG_(show_all_errors)();
 
-   VG_TDICT_CALL(tool_fini, 0/*exitcode*/);
-
    if (VG_(clo_xml)) {
       VG_(message)(Vg_UserMsg, "");
       VG_(message)(Vg_UserMsg, "</valgrindoutput>");
