Index: memcheck/mc_include.h
===================================================================
--- memcheck/mc_include.h	(revision 10888)
+++ memcheck/mc_include.h	(working copy)
@@ -43,7 +43,7 @@
 /*------------------------------------------------------------*/
 
 /* We want at least a 16B redzone on client heap blocks for Memcheck */
-#define MC_MALLOC_REDZONE_SZB    16
+#define MC_MALLOC_REDZONE_SZB    64
 
 /* For malloc()/new/new[] vs. free()/delete/delete[] mismatch checking. */
 typedef
