# See https://bugs.kde.org/show_bug.cgi?id=210481
Index: guest_amd64_toIR.c
===================================================================
--- VEX/priv/guest_amd64_toIR.c	(Revision 1924)
+++ VEX/priv/guest_amd64_toIR.c	(Arbeitskopie)
@@ -14597,7 +14597,7 @@
       /* There is no encoding for 32-bit pop in 64-bit mode.
          So sz==4 actually means sz==8. */
       if (haveF2orF3(pfx)) goto decode_failure;
-      vassert(sz == 2 || sz == 4);
+      vassert(sz == 2 || sz == 4 || sz == 8);
       if (sz == 4) sz = 8;
       if (sz != 8) goto decode_failure; // until we know a sz==2 test case exists
 
