diff --git a/third_party/tlslite/tlslite/utils/RSAKey.py b/third_party/tlslite/tlslite/utils/RSAKey.py
index 37c292d..1b91742 100644
--- a/third_party/tlslite/tlslite/utils/RSAKey.py
+++ b/third_party/tlslite/tlslite/utils/RSAKey.py
@@ -117,7 +117,7 @@ class RSAKey:
         if m >= self.n:
             raise ValueError()
         c = self._rawPrivateKeyOp(m)
-        sigBytes = numberToBytes(c)
+        sigBytes = numberToBytes(c, numBytes(self.n))
         return sigBytes
 
     def verify(self, sigBytes, bytes):
diff --git a/third_party/tlslite/tlslite/utils/cryptomath.py b/third_party/tlslite/tlslite/utils/cryptomath.py
index 385095d..86da25e 100644
--- a/third_party/tlslite/tlslite/utils/cryptomath.py
+++ b/third_party/tlslite/tlslite/utils/cryptomath.py
@@ -129,8 +129,9 @@ def bytesToNumber(bytes):
         multiplier *= 256
     return total
 
-def numberToBytes(n):
-    howManyBytes = numBytes(n)
+def numberToBytes(n, howManyBytes=None):
+    if howManyBytes == None:
+      howManyBytes = numBytes(n)
     bytes = createByteArrayZeros(howManyBytes)
     for count in range(howManyBytes-1, -1, -1):
         bytes[count] = int(n % 256)
