diff -Naur gmock-1.5.0/Makefile gmock-1.5.0.nacl/Makefile
--- gmock-1.5.0/Makefile	1969-12-31 17:00:00.000000000 -0700
+++ gmock-1.5.0.nacl/Makefile	2010-07-07 14:28:27.000000000 -0600
@@ -0,0 +1,25 @@
+CXXFLAGS = -O0 -g
+INCLUDE = -Iinclude \
+          -I. \
+          -I$(NACL_TOOLCHAIN_ROOT)/nacl/usr/include \
+          -I$(NACL_TOOLCHAIN_ROOT)/nacl64/usr/include
+LIB_GMOCK = libgmock.a
+OBJ_DIR = obj
+
+OBJ_FILES = gmock-all.o \
+            gmock_main.o
+
+all: $(LIB_GMOCK)
+	@echo "Making libgmock"
+
+clean:
+	rm -rf $(OBJ_DIR) $(LIB_GMOCK)
+
+$(OBJ_DIR):
+	mkdir $(OBJ_DIR)
+
+$(OBJ_FILES): %.o: src/%.cc $(OBJ_DIR)
+	$(CXX) -c $(INCLUDE) $(CXXFLAGS) $< -o $(OBJ_DIR)/$@
+
+$(LIB_GMOCK): $(OBJ_DIR) $(OBJ_FILES)
+	cd $(OBJ_DIR) && $(AR) rcs ../$(LIB_GMOCK) $(OBJ_FILES)
