diff --git a/third_party/unrar/src/archive.hpp b/third_party/unrar/src/archive.hpp
index c2dec52041a9..9c1456830413 100644
--- a/third_party/unrar/src/archive.hpp
+++ b/third_party/unrar/src/archive.hpp
@@ -94,7 +94,7 @@ class Archive:public File
     void AddSubData(byte *SrcData,uint64 DataSize,File *SrcFile,
          const wchar *Name,uint Flags);
     bool ReadSubData(Array<byte> *UnpData,File *DestFile);
-    HEADER_TYPE GetHeaderType() {return CurHeaderType;};
+    HEADER_TYPE GetHeaderType() {return CurHeaderType;}
     RAROptions* GetRAROptions() {return Cmd;}
     void SetSilentOpen(bool Mode) {SilentOpen=Mode;}
 #if 0
diff --git a/third_party/unrar/src/errhnd.hpp b/third_party/unrar/src/errhnd.hpp
index 3dff399b16bb..b3f728e48122 100644
--- a/third_party/unrar/src/errhnd.hpp
+++ b/third_party/unrar/src/errhnd.hpp
@@ -57,7 +57,7 @@ class ErrorHandler
     uint GetErrorCount() {return ErrCount;}
     void SetSignalHandlers(bool Enable);
     void Throw(RAR_EXIT Code);
-    void SetSilent(bool Mode) {Silent=Mode;};
+    void SetSilent(bool Mode) {Silent=Mode;}
     bool GetSysErrMsg(wchar *Msg,size_t Size);
     void SysErrMsg();
     int GetSystemErrorCode();
diff --git a/third_party/unrar/src/file.hpp b/third_party/unrar/src/file.hpp
index 4b3b637ad832..9cc7807b0fde 100644
--- a/third_party/unrar/src/file.hpp
+++ b/third_party/unrar/src/file.hpp
@@ -105,7 +105,7 @@ class File
     void SetCloseFileTime(RarTime *ftm,RarTime *fta=NULL);
     static void SetCloseFileTimeByName(const wchar *Name,RarTime *ftm,RarTime *fta);
     void GetOpenFileTime(RarTime *ft);
-    virtual bool IsOpened() {return hFile!=FILE_BAD_HANDLE;}; // 'virtual' for MultiFile class.
+    virtual bool IsOpened() {return hFile!=FILE_BAD_HANDLE;} // 'virtual' for MultiFile class.
     int64 FileLength();
     void SetHandleType(FILE_HANDLETYPE Type) {HandleType=Type;}
     FILE_HANDLETYPE GetHandleType() {return HandleType;}
diff --git a/third_party/unrar/src/scantree.hpp b/third_party/unrar/src/scantree.hpp
index 73354ab9429f..9edcb24527aa 100644
--- a/third_party/unrar/src/scantree.hpp
+++ b/third_party/unrar/src/scantree.hpp
@@ -66,8 +66,8 @@ class ScanTree
     ScanTree(StringList *FileMasks,RECURSE_MODE Recurse,bool GetLinks,SCAN_DIRS GetDirs);
     ~ScanTree();
     SCAN_CODE GetNext(FindData *FindData);
-    size_t GetSpecPathLength() {return SpecPathLength;};
-    int GetErrors() {return Errors;};
+    size_t GetSpecPathLength() {return SpecPathLength;}
+    int GetErrors() {return Errors;}
     void SetErrArcName(const wchar *Name) {wcsncpyz(ErrArcName,Name,ASIZE(ErrArcName));}
     void SetCommandData(CommandData *Cmd) {ScanTree::Cmd=Cmd;}
     void SetErrDirList(StringList *List,Array<uint> *Lengths)
diff --git a/third_party/unrar/src/strlist.hpp b/third_party/unrar/src/strlist.hpp
index 55108dd9b10a..2c16b68f18eb 100644
--- a/third_party/unrar/src/strlist.hpp
+++ b/third_party/unrar/src/strlist.hpp
@@ -23,7 +23,7 @@ class StringList
     wchar* GetString();
     bool GetString(wchar **Str);
     void Rewind();
-    size_t ItemsCount() {return StringsCount;};
+    size_t ItemsCount() {return StringsCount;}
     size_t GetCharCount() {return StringData.Size();}
     bool Search(const wchar *Str,bool CaseSensitive);
     void SavePosition();
diff --git a/third_party/unrar/src/suballoc.hpp b/third_party/unrar/src/suballoc.hpp
index 17af204eb6c1..98ba2dea73ff 100644
--- a/third_party/unrar/src/suballoc.hpp
+++ b/third_party/unrar/src/suballoc.hpp
@@ -79,7 +79,7 @@ class SubAllocator
     inline void* ExpandUnits(void* ptr,int OldNU);
     inline void* ShrinkUnits(void* ptr,int OldNU,int NewNU);
     inline void  FreeUnits(void* ptr,int OldNU);
-    long GetAllocatedMemory() {return(SubAllocatorSize);};
+    long GetAllocatedMemory() {return(SubAllocatorSize);}
 
     byte *pText, *UnitsStart,*HeapEnd,*FakeUnitsStart;
 };
