diff --git a/source/i18n/formatted_string_builder.h b/source/i18n/formatted_string_builder.h
index 4567dc1d..51690d37 100644
--- a/source/i18n/formatted_string_builder.h
+++ b/source/i18n/formatted_string_builder.h
@@ -224,8 +224,10 @@ static_assert(
 
 constexpr FormattedStringBuilder::Field::Field(uint8_t category, uint8_t field)
     : bits((
+#if 0 // /ICU-21081
         U_ASSERT(category <= 0xf),
         U_ASSERT(field <= 0xf),
+#endif
         static_cast<uint8_t>((category << 4) | field)
     )) {}
 
