diff --git a/src/main/java/com/android/tools/r8/utils/InternalOptions.java b/src/main/java/com/android/tools/r8/utils/InternalOptions.java
index 595ac28da..e96f06364 100644
--- a/src/main/java/com/android/tools/r8/utils/InternalOptions.java
+++ b/src/main/java/com/android/tools/r8/utils/InternalOptions.java
@@ -194,8 +194,8 @@ public class InternalOptions {
   public boolean enableDynamicTypeOptimization = true;
   public boolean enableFieldBitAccessAnalysis =
       System.getProperty("com.android.tools.r8.fieldBitAccessAnalysis") != null;
-  public boolean enableHorizontalClassMerging = true;
-  public boolean enableVerticalClassMerging = true;
+  public boolean enableHorizontalClassMerging = false;
+  public boolean enableVerticalClassMerging = false;
   public boolean enableArgumentRemoval = true;
   public boolean enableUnusedArgumentRemoval = true;
   public boolean enableUnusedInterfaceRemoval = true;
@@ -919,7 +919,7 @@ public class InternalOptions {
     public static final String CLASS_NAME = "com.android.tools.r8.GeneratedOutlineSupport";
     public static final String METHOD_PREFIX = "outline";
 
-    public boolean enabled = true;
+    public boolean enabled = System.getProperty("com.android.tools.r8.disableOutlining") == null;
     public int minSize = 3;
     public int maxSize = 99;
     public int threshold = 20;
