diff -Naur PRISTINE/ImageMagick-6.4.5/config/config.sub MODIFIED/ImageMagick-6.4.5/config/config.sub
--- PRISTINE/ImageMagick-6.4.5/config/config.sub	2008-10-29 07:41:37.000000000 -0700
+++ MODIFIED/ImageMagick-6.4.5/config/config.sub	2008-11-17 11:36:24.297614000 -0800
@@ -116,6 +116,10 @@
     exit 1;;
 esac
 
+case "$1" in
+nacl)	echo i386-google-nacl; exit;;
+esac
+
 # Separate what the user gave into CPU-COMPANY and OS or KERNEL-OS (if any).
 # Here we must recognize all the valid KERNEL-OS combinations.
 maybe_os=`echo $1 | sed 's/^\(.*\)-\([^-]*-[^-]*\)$/\2/'`
diff -Naur PRISTINE/ImageMagick-6.4.5/Magick++/lib/Geometry.cpp MODIFIED/ImageMagick-6.4.5/Magick++/lib/Geometry.cpp
--- PRISTINE/ImageMagick-6.4.5/Magick++/lib/Geometry.cpp	2008-07-28 06:36:47.000000000 -0700
+++ MODIFIED/ImageMagick-6.4.5/Magick++/lib/Geometry.cpp	2008-11-17 13:28:24.244935000 -0800
@@ -10,7 +10,7 @@
 #include "Magick++/Include.h"
 #include <string>
 #include <ctype.h> // for isdigit
-#if !defined(__WINDOWS__)
+#if !defined(__WINDOWS__) && MAGICKCORE_HAVE_STRINGS_H
 #include <strings.h>
 #endif
 
diff -Naur PRISTINE/ImageMagick-6.4.5/Magick++/lib/Image.cpp MODIFIED/ImageMagick-6.4.5/Magick++/lib/Image.cpp
--- PRISTINE/ImageMagick-6.4.5/Magick++/lib/Image.cpp	2008-11-14 06:09:59.000000000 -0800
+++ MODIFIED/ImageMagick-6.4.5/Magick++/lib/Image.cpp	2008-11-17 13:41:44.524857000 -0800
@@ -13,7 +13,7 @@
 #include <string.h>
 #include <errno.h>
 #include <math.h>
-#if !defined(__WINDOWS__)
+#if !defined(__WINDOWS__) && MAGICKCORE_HAVE_STRINGS_H
 #include <strings.h>
 #endif
 
