diff --git a/third_party/re2/re2/parse.cc b/third_party/re2/re2/parse.cc
index 0cf4ab4..6423fe9 100644
--- a/third_party/re2/re2/parse.cc
+++ b/third_party/re2/re2/parse.cc
@@ -16,6 +16,8 @@
 // and recognizes the Perl escape sequences \d, \s, \w, \D, \S, and \W.
 // See regexp.h for rationale.
 
+#include <ctype.h>
+
 #include "util/util.h"
 #include "re2/regexp.h"
 #include "re2/stringpiece.h"
diff --git a/third_party/re2/re2/re2.cc b/third_party/re2/re2/re2.cc
index 989add6..78978f1 100644
--- a/third_party/re2/re2/re2.cc
+++ b/third_party/re2/re2/re2.cc
@@ -9,6 +9,8 @@
 
 #include "re2/re2.h"
 
+#include <ctype.h>
+
 #include <stdio.h>
 #include <string>
 #ifdef WIN32
diff --git a/third_party/re2/util/util.h b/third_party/re2/util/util.h
index dab7e16..11b5f4a 100644
--- a/third_party/re2/util/util.h
+++ b/third_party/re2/util/util.h
@@ -28,6 +28,7 @@
 #include <utility>
 #include <set>
 
+#include "build/build_config.h"
 #include "base/third_party/dynamic_annotations/dynamic_annotations.h"
 
 // Use std names.
@@ -44,7 +45,7 @@ using std::sort;
 using std::swap;
 using std::make_pair;
 
-#if defined(__GNUC__) && !defined(USE_CXX0X)
+#if defined(__GNUC__) && !defined(USE_CXX0X) && !defined(OS_ANDROID)
 
 #include <tr1/unordered_set>
 using std::tr1::unordered_set;
@@ -52,7 +53,7 @@ using std::tr1::unordered_set;
 #else
 
 #include <unordered_set>
-#ifdef WIN32
+#if defined(WIN32) || defined(OS_ANDROID)
 using std::tr1::unordered_set;
 #else
 using std::unordered_set;
