diff --git a/third_party/re2/re2/unicode.py b/third_party/re2/re2/unicode.py
old mode 100755
new mode 100644
diff --git a/third_party/re2/testinstall.cc b/third_party/re2/testinstall.cc
deleted file mode 100644
index 40b7a8a..0000000
--- a/third_party/re2/testinstall.cc
+++ /dev/null
@@ -1,20 +0,0 @@
-#include <re2/re2.h>
-#include <re2/filtered_re2.h>
-#include <stdio.h>
-
-using namespace re2;
-
-int main(void) {
-	FilteredRE2 f;
-	int id;
-	f.Add("a.*b.*c", RE2::DefaultOptions, &id);
-	vector<string> v;
-	f.Compile(&v);
-
-	if(RE2::FullMatch("axbyc", "a.*b.*c")) {
-		printf("PASS\n");
-		return 0;
-	}
-	printf("FAIL\n");
-	return 2;
-}
