diff --git a/third_party/re2/util/sparse_array.h b/third_party/re2/util/sparse_array.h
index 4ee5c94..7bc3a86 100644
--- a/third_party/re2/util/sparse_array.h
+++ b/third_party/re2/util/sparse_array.h
@@ -273,13 +273,13 @@ void SparseArray<Value>::resize(int new_max_size) {
     int* a = new int[new_max_size];
     if (sparse_to_dense_) {
       memmove(a, sparse_to_dense_, max_size_*sizeof a[0]);
-      // Don't need to zero the memory but appease Valgrind.
-      if (valgrind_) {
-        for (int i = max_size_; i < new_max_size; i++)
-          a[i] = 0xababababU;
-      }
       delete[] sparse_to_dense_;
     }
+    // Don't need to zero the memory but appease Valgrind.
+    if (valgrind_) {
+      for (int i = max_size_; i < new_max_size; i++)
+        a[i] = 0xababababU;
+    }
     sparse_to_dense_ = a;
 
     dense_.resize(new_max_size);
