diff --git a/components-chromium/font-roboto/roboto.html b/components-chromium/font-roboto/roboto.html
index 7a24999..4eefcba 100644
--- a/components-chromium/font-roboto/roboto.html
+++ b/components-chromium/font-roboto/roboto.html
@@ -7,5 +7,4 @@ The complete set of contributors may be found at http://polymer.github.io/CONTRI
 Code distributed by Google as part of the polymer project is also
 subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
 -->
-<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:400,300,300italic,400italic,500,500italic,700,700italic">
-<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto+Mono:400,700">
+<link rel="stylesheet" href="chrome://resources/css/roboto.css">
diff --git a/components-chromium/paper-input/paper-textarea-extracted.js b/components-chromium/paper-input/paper-textarea-extracted.js
index 78a2a9ec621c..8b776dd790ea 100644
--- a/components-chromium/paper-input/paper-textarea-extracted.js
+++ b/components-chromium/paper-input/paper-textarea-extracted.js
@@ -43,8 +43,8 @@ Polymer({
       }
     },
 
-    _ariaLabelledByChanged: function(ariaLabelledBy) {
-      this.$.input.textarea.setAttribute('aria-labelledby', ariaLabelledBy);
+    _ariaLabelledByChanged: function() {
+      this.$.input.textarea.setAttribute('aria-label', this.label);
     },
 
     _ariaDescribedByChanged: function(ariaDescribedBy) {
diff --git a/components-chromium/paper-icon-button/paper-icon-button-light-extracted.js b/components-chromium/paper-icon-button/paper-icon-button-light-extracted.js
index bac589c7274c..bc25f54320fb 100644
--- a/components-chromium/paper-icon-button/paper-icon-button-light-extracted.js
+++ b/components-chromium/paper-icon-button/paper-icon-button-light-extracted.js
@@ -14,11 +14,11 @@ Polymer({
       },

       _rippleDown: function() {
-        this.getRipple().downAction();
+        this.getRipple().uiDownAction();
       },

       _rippleUp: function() {
-        this.getRipple().upAction();
+        this.getRipple().uiUpAction();
       },

       /**
