diff -Naur slider/default.css slider_mod/default.css
--- slider/default.css	2009-06-01 17:25:14.327738000 -0700
+++ slider_mod/default.css	2009-06-01 17:25:25.768785000 -0700
@@ -111,5 +111,6 @@
 	font-weight: bold;
 	border: 0;
 	cursor: default;
+	updatefnc: undefined;
 }
 
diff -Naur slider/slider.js slider_mod/slider.js
--- slider/slider.js	2009-05-29 15:53:56.419969000 -0700
+++ slider_mod/slider.js	2009-05-29 15:54:25.302442000 -0700
@@ -104,6 +104,10 @@
 		var v = Math.round((sliderPos * carpeslider.scale + carpeslider.from) * // calculate display value
 			Math.pow(10, carpedisplay.decimals)) / Math.pow(10, carpedisplay.decimals);
 		carpedisplay.value = v; // put the new value in the slider display element
+		var updatefnc = carpeslider.getAttribute('updatefnc');
+		if (updatefnc != undefined) {
+			eval(updatefnc+'(v)');
+		}
 		return false;
 	}
 	return
