--- jpeg-6b/config.sub	1998-03-25 07:25:39.000000000 -0800
+++ jpeg-6b.new/config.sub	2009-04-28 16:54:16.738799000 -0700
@@ -408,6 +408,10 @@
 	mips3*)
 		basic_machine=`echo $basic_machine | sed -e 's/mips3/mips64/'`-unknown
 		;;
+	nacl*)
+		basic_machine=i686-pc
+		os=-nacl
+		;;
 	ncr3000)
 		basic_machine=i486-ncr
 		os=-sysv4
@@ -716,7 +720,7 @@
 	      | -ptx* | -coff* | -ecoff* | -winnt* | -domain* | -vsta* \
 	      | -udi* | -eabi* | -lites* | -ieee* | -go32* | -aux* \
 	      | -cygwin32* | -pe* | -psos* | -moss* | -proelf* | -rtems* \
-	      | -mingw32* | -linux-gnu* | -uxpv*)
+	      | -mingw32* | -linux-gnu* | -uxpv* | -nacl*)
 	# Remember, each alternative MUST END IN *, to match a version number.
 		;;
 	-linux*)
