diff --git a/third_party/protobuf/src/google/protobuf/implicit_weak_message.cc b/third_party/protobuf/src/google/protobuf/implicit_weak_message.cc
index 7a1d4446bf32..5189efba35e9 100644
--- a/third_party/protobuf/src/google/protobuf/implicit_weak_message.cc
+++ b/third_party/protobuf/src/google/protobuf/implicit_weak_message.cc
@@ -33,6 +33,7 @@
 #include <google/protobuf/stubs/once.h>
 #include <google/protobuf/io/zero_copy_stream_impl_lite.h>
 #include <google/protobuf/wire_format_lite.h>
+#include <google/protobuf/wire_format_lite_inl.h>
 
 namespace google {
 namespace protobuf {
