From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Victor Costan <pwnall@chromium.org>
Date: Sun, 10 Feb 2019 15:18:43 -0800
Subject: [PATCH 06/10] Fix dbfuzz2.c compilation errors on Windows.

---
 third_party/sqlite/src/test/dbfuzz2.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/third_party/sqlite/src/test/dbfuzz2.c b/third_party/sqlite/src/test/dbfuzz2.c
index c964ad4f018f..dd1ed061717b 100644
--- a/third_party/sqlite/src/test/dbfuzz2.c
+++ b/third_party/sqlite/src/test/dbfuzz2.c
@@ -43,8 +43,10 @@
 #include <stdarg.h>
 #include <ctype.h>
 #include <stdint.h>
+#ifndef _WIN32
 #include <sys/time.h>
 #include <sys/resource.h>
+#endif
 #include "sqlite3.h"
 
 /*
@@ -261,6 +263,7 @@ int LLVMFuzzerInitialize(int *pArgc, char ***pArgv){
         szMax = strtol(argv[++i], 0, 0);
         continue;
       }
+#ifndef _WIN32
       if( strcmp(z,"max-stack")==0
        || strcmp(z,"max-data")==0
        || strcmp(z,"max-as")==0
@@ -291,6 +294,7 @@ int LLVMFuzzerInitialize(int *pArgc, char ***pArgv){
                zType, (int)x.rlim_cur, (int)y.rlim_cur);
         continue;
       }
+#endif
     }
     argv[j++] = argv[i];
   }
-- 
2.21.0.225.g810b269d1ac-goog

