diff --git a/third_party/document_image_extractor/third_party/closure-library/closure/goog/dom/dom.js b/third_party/document_image_extractor/third_party/closure-library/closure/goog/dom/dom.js
index 9b9ec9e..2d0cdd0 100644
--- a/third_party/document_image_extractor/third_party/closure-library/closure/goog/dom/dom.js
+++ b/third_party/document_image_extractor/third_party/closure-library/closure/goog/dom/dom.js
@@ -249,7 +249,7 @@ goog.dom.getElementByClass = function(className, opt_el) {
  * @param {string} className the name of the class to look for.
  * @param {!Element|!Document=} opt_root Optional element or document to look
  *     in.
- * @return {!Element} The first item with the class name provided.
+ * @return {!Element|null} The first item with the class name provided.
  * @throws {goog.asserts.AssertionError} Thrown if no element is found.
  */
 goog.dom.getRequiredElementByClass = function(className, opt_root) {
@@ -2340,7 +2340,7 @@ goog.dom.DomHelper.prototype.getElementByClass = function(className, opt_el) {
  * @param {string} className the name of the class to look for.
  * @param {(!Element|!Document)=} opt_root Optional element or document to look
  *     in.
- * @return {!Element} The first item found with the class name provided.
+ * @return {!Element|null} The first item found with the class name provided.
  * @throws {goog.asserts.AssertionError} Thrown if no element is found.
  */
 goog.dom.DomHelper.prototype.getRequiredElementByClass = function(className,
diff --git a/third_party/document_image_extractor/third_party/closure-library/closure/goog/style/style.js b/third_party/document_image_extractor/third_party/closure-library/closure/goog/style/style.js
index c9179b2f..9ed0efa 100644
--- a/third_party/document_image_extractor/third_party/closure-library/closure/goog/style/style.js
+++ b/third_party/document_image_extractor/third_party/closure-library/closure/goog/style/style.js
@@ -38,8 +38,8 @@ goog.require('goog.object');
 goog.require('goog.string');
 goog.require('goog.userAgent');
 
-goog.forwardDeclare('goog.events.BrowserEvent');
-goog.forwardDeclare('goog.events.Event');
+goog.require('goog.events.BrowserEvent');
+goog.require('goog.events.Event');
 
 
 /**
@@ -438,7 +438,7 @@ goog.style.getViewportPageOffset = function(doc) {
  * If the element is not in the DOM, the result is undefined, and an error may
  * be thrown depending on user agent.
  *
- * @param {!Element} el The element whose bounding rectangle is being queried.
+ * @param {!Element|null} el The element whose bounding rectangle is being queried.
  * @return {Object} A native bounding rectangle with numerical left, top,
  *     right, and bottom.  Reported by Firefox to be of object type ClientRect.
  * @private
@@ -819,7 +819,7 @@ goog.style.getRelativePosition = function(a, b) {
 /**
  * Returns the position of the event or the element's border box relative to
  * the client viewport.
- * @param {!Element} el Element whose position to get.
+ * @param {!Element|null} el Element whose position to get.
  * @return {!goog.math.Coordinate} The position.
  * @private
  */
diff --git a/third_party/document_image_extractor/third_party/closure-library/closure/goog/events/eventhandler.js b/third_party/document_image_extractor/third_party/closure-library/closure/goog/events/eventhandler.js
index 16b1ad0..3551d10 100644
--- a/third_party/document_image_extractor/third_party/closure-library/closure/goog/events/eventhandler.js
+++ b/third_party/document_image_extractor/third_party/closure-library/closure/goog/events/eventhandler.js
@@ -61,7 +61,7 @@ goog.require('goog.Disposable');
 goog.require('goog.events');
 goog.require('goog.object');
 
-goog.forwardDeclare('goog.events.EventWrapper');
+goog.require('goog.events.EventWrapper');
 
 
 
diff --git a/third_party/document_image_extractor/third_party/closure-library/closure/goog/events/events.js b/third_party/document_image_extractor/third_party/closure-library/closure/goog/events/events.js
index 39cc405..7687b5c 100644
--- a/third_party/document_image_extractor/third_party/closure-library/closure/goog/events/events.js
+++ b/third_party/document_image_extractor/third_party/closure-library/closure/goog/events/events.js
@@ -63,8 +63,8 @@ goog.require('goog.events.BrowserFeature');
 goog.require('goog.events.Listenable');
 goog.require('goog.events.ListenerMap');
 
-goog.forwardDeclare('goog.debug.ErrorHandler');
-goog.forwardDeclare('goog.events.EventWrapper');
+goog.require('goog.debug.ErrorHandler');
+goog.require('goog.events.EventWrapper');
 
 
 /**
diff --git a/third_party/document_image_extractor/third_party/closure-library/closure/goog/ui/component.js b/third_party/document_image_extractor/third_party/closure-library/closure/goog/ui/component.js
index b8806fd..0ee9b9f 100644
--- a/third_party/document_image_extractor/third_party/closure-library/closure/goog/ui/component.js
+++ b/third_party/document_image_extractor/third_party/closure-library/closure/goog/ui/component.js
@@ -474,7 +474,7 @@ goog.ui.Component.prototype.getElement = function() {
  * it assumes that the element exists (i.e. the component has been
  * rendered/decorated) and will cause an assertion error otherwise (if
  * assertion is enabled).
- * @return {!Element} The element for the component.
+ * @return {!Element|null} The element for the component.
  */
 goog.ui.Component.prototype.getElementStrict = function() {
   var el = this.element_;
@@ -530,7 +530,7 @@ goog.ui.Component.prototype.getElementByClass = function(className) {
  * element to be present in the dom thus returning a required value. Otherwise,
  * will assert.
  * @param {string} className The name of the class to look for.
- * @return {!Element} The first item with the class name provided.
+ * @return {!Element|null} The first item with the class name provided.
  */
 goog.ui.Component.prototype.getRequiredElementByClass = function(className) {
   var el = this.getElementByClass(className);
