diff --git a/libavutil/attributes.h b/libavutil/attributes.h
index 0a6fda1..9a801d5 100644
--- a/libavutil/attributes.h
+++ b/libavutil/attributes.h
@@ -89,11 +89,12 @@
 #endif
 
 #ifndef attribute_deprecated
-#if AV_GCC_VERSION_AT_LEAST(3,1)
-#    define attribute_deprecated __attribute__((deprecated))
-#else
+// TODO(dalecurtis): Revert once http://crbug.com/112673 is fixed.
+//#if AV_GCC_VERSION_AT_LEAST(3,1)
+//#    define attribute_deprecated __attribute__((deprecated))
+//#else
 #    define attribute_deprecated
-#endif
+//#endif
 #endif
 
 /**
