diff -wurp -N orig/libavcodec/mpegaudiodec.c ffmpeg/libavcodec/mpegaudiodec.c
--- orig/libavcodec/mpegaudiodec.c	2011-07-12 20:50:08.712944821 -0700
+++ ffmpeg/libavcodec/mpegaudiodec.c	2011-07-12 20:50:08.782964340 -0700
@@ -1704,6 +1704,7 @@ static int mp_decode_frame(MPADecodeCont
     OUT_INT *samples_ptr;
 
     init_get_bits(&s->gb, buf + HEADER_SIZE, (buf_size - HEADER_SIZE)*8);
+    s->gb.buffer_enforcing = 1;
 
     /* skip error protection field */
     if (s->error_protection)
