diff -wurp -N orig/libavcodec/avcodec.h ffmpeg/libavcodec/avcodec.h
--- orig/libavcodec/avcodec.h	2011-07-12 20:50:13.692979736 -0700
+++ ffmpeg/libavcodec/avcodec.h	2011-07-12 20:50:13.755019764 -0700
@@ -526,7 +526,8 @@ enum AVChromaLocation{
 /**
  * LPC analysis type
  */
-attribute_deprecated enum AVLPCType {
+//attribute_deprecated
+enum AVLPCType {
     AV_LPC_TYPE_DEFAULT     = -1, ///< use the codec default LPC type
     AV_LPC_TYPE_NONE        =  0, ///< do not use LPC prediction or use all zero coefficients
     AV_LPC_TYPE_FIXED       =  1, ///< fixed LPC coefficients
@@ -2795,14 +2796,16 @@ typedef struct AVCodecContext {
      * - encoding: Set by user
      * - decoding: unused
      */
-    attribute_deprecated enum AVLPCType lpc_type;
+    //attribute_deprecated
+    enum AVLPCType lpc_type;
 
     /**
      * Number of passes to use for Cholesky factorization during LPC analysis
      * - encoding: Set by user
      * - decoding: unused
      */
-    attribute_deprecated int lpc_passes;
+    //attribute_deprecated
+    int lpc_passes;
 #endif
 
     /**
diff -wurp -N orig/libavformat/avio.h ffmpeg/libavformat/avio.h
--- orig/libavformat/avio.h	2011-07-12 20:50:13.703263949 -0700
+++ ffmpeg/libavformat/avio.h	2011-07-12 20:50:13.755019764 -0700
@@ -209,7 +209,8 @@ URLProtocol *av_protocol_next(URLProtoco
  *
  * @param size the size of the URLProtocol struct referenced
  */
-attribute_deprecated int av_register_protocol2(URLProtocol *protocol, int size);
+//attribute_deprecated
+int av_register_protocol2(URLProtocol *protocol, int size);
 /**
  * @}
  */
diff -wurp -N orig/libavutil/samplefmt.h ffmpeg/libavutil/samplefmt.h
--- orig/libavutil/samplefmt.h	2011-07-12 20:50:13.713031456 -0700
+++ ffmpeg/libavutil/samplefmt.h	2011-07-12 20:50:13.755019764 -0700
@@ -64,7 +64,7 @@ char *av_get_sample_fmt_string(char *buf
 /**
  * @deprecated Use av_get_bytes_per_sample() instead.
  */
-attribute_deprecated
+//attribute_deprecated
 int av_get_bits_per_sample_fmt(enum AVSampleFormat sample_fmt);
 #endif

diff -wurp -N orig/libavcodec/avcodec.h ffmpeg/libavcodec/avcodec.h
--- orig/libavcodec/avcodec.h	2011-07-12 21:38:32.792973129 -0700
+++ffmpeg/libavcodec/avcodec.h	2011-07-14 12:01:26.002994746 -0700
@@ -3625,7 +3625,7 @@ int avcodec_get_context_defaults3(AVCode
  *
  * @deprecated use avcodec_alloc_context3()
  */
-attribute_deprecated
+//attribute_deprecated
 AVCodecContext *avcodec_alloc_context(void);
 
 /** THIS FUNCTION IS NOT YET PART OF THE PUBLIC API!
@@ -3756,7 +3756,7 @@ int avcodec_default_execute2(AVCodecCont
  *
  * @deprecated use avcodec_open2
  */
-attribute_deprecated
+//attribute_deprecated
 int avcodec_open(AVCodecContext *avctx, AVCodec *codec);
 #endif
 
diff -wurp -N orig/libavformat/avformat.h ffmpeg/libavformat/avformat.h
--- orig/libavformat/avformat.h	2011-07-12 15:42:11.000000000 -0700
+++ ffmpeg/libavformat/avformat.h	2011-07-14 12:00:55.842983958 -0700
@@ -1107,7 +1107,8 @@ attribute_deprecated int av_open_input_s
  *
  * @deprecated use avformat_open_input instead.
  */
-attribute_deprecated int av_open_input_file(AVFormatContext **ic_ptr, const char *filename,
+//attribute_deprecated
+int av_open_input_file(AVFormatContext **ic_ptr, const char *filename,
                        AVInputFormat *fmt,
                        int buf_size,
                        AVFormatParameters *ap);

