diff --git a/libavcodec/x86/cabac.h b/libavcodec/x86/cabac.h
index 2bb0be6..345b510 100644
--- a/libavcodec/x86/cabac.h
+++ b/libavcodec/x86/cabac.h
@@ -102,6 +102,10 @@ static av_always_inline int get_cabac_inline_x86(CABACContext *c,
 }
 #endif /* HAVE_7REGS && !defined(BROKEN_RELOCATIONS) */

+// TODO(dalecurtis): Required to get Mac ia32 builds compiling with
+// -fno-omit-frame-pointer, which is required for accurate stack traces.  See
+// http://crbug.com/115170.
+#if !(defined(__i386) && defined(__clang__) && defined(__APPLE__))
 #define get_cabac_bypass_sign get_cabac_bypass_sign_x86
 static av_always_inline int get_cabac_bypass_sign_x86(CABACContext *c, int val)
 {
@@ -136,5 +140,6 @@ static av_always_inline int get_cabac_bypass_sign_x86(CABACContext *c, int val)
     );
     return val;
 }
+#endif /* !(defined(__i386) && defined(__clang__) && defined(__APPLE__)) */

 #endif /* AVCODEC_X86_CABAC_H */
--
1.7.7.3

