diff --git a/libavcodec/avcodec.h b/libavcodec/avcodec.h
index e5b0c2b..8295f35 100644
--- a/libavcodec/avcodec.h
+++ b/libavcodec/avcodec.h
@@ -1072,9 +1072,11 @@ typedef struct AVPacket {
      */
     int   duration;
 #if FF_API_DESTRUCT_PACKET
-    attribute_deprecated
+    // TODO(wolenetz) Switch to refcounted buffers. See http://crbug.com/236611
+    // attribute_deprecated
     void  (*destruct)(struct AVPacket *);
-    attribute_deprecated
+    // TODO(wolenetz) Switch to refcounted buffers. See http://crbug.com/236611
+    // attribute_deprecated
     void  *priv;
 #endif
     int64_t pos;                            ///< byte position in stream, -1 if unknown
@@ -2006,7 +2008,8 @@ typedef struct AVCodecContext {
      *
      * @deprecated use get_buffer2()
      */
-    attribute_deprecated
+    // TODO(wolenetz) Switch to refcounted buffers. See http://crbug.com/236611
+    // attribute_deprecated
     int (*get_buffer)(struct AVCodecContext *c, AVFrame *pic);
 
     /**
@@ -2020,7 +2023,8 @@ typedef struct AVCodecContext {
      *
      * @deprecated custom freeing callbacks should be set from get_buffer2()
      */
-    attribute_deprecated
+    // TODO(wolenetz) Switch to refcounted buffers. See http://crbug.com/236611
+    // attribute_deprecated
     void (*release_buffer)(struct AVCodecContext *c, AVFrame *pic);
 
     /**
@@ -3389,7 +3393,8 @@ void avsubtitle_free(AVSubtitle *sub);
  * Default packet destructor.
  * @deprecated use the AVBuffer API instead
  */
-attribute_deprecated
+// TODO(wolenetz) Switch to refcounted buffers. See http://crbug.com/236611
+// attribute_deprecated
 void av_destruct_packet(AVPacket *pkt);
 #endif
 
diff --git a/libavutil/frame.h b/libavutil/frame.h
index 39a664f..a24c299 100644
--- a/libavutil/frame.h
+++ b/libavutil/frame.h
@@ -144,7 +144,8 @@ typedef struct AVFrame {
     enum AVPictureType pict_type;
 
 #if FF_API_AVFRAME_LAVC
-    attribute_deprecated
+    // TODO(wolenetz) Switch to refcounted buffers. See http://crbug.com/236611
+    // attribute_deprecated
     uint8_t *base[AV_NUM_DATA_POINTERS];
 #endif
 
@@ -254,7 +255,8 @@ typedef struct AVFrame {
     uint64_t error[AV_NUM_DATA_POINTERS];
 
 #if FF_API_AVFRAME_LAVC
-    attribute_deprecated
+    // TODO(wolenetz) Switch to refcounted buffers. See http://crbug.com/236611
+    // attribute_deprecated
     int type;
 #endif
 
